clang-tidy: Fix `performance-*move*` warnings in headers

hebasto opened the pull request:
Split from bitcoin/bitcoin#26705 as was requested in https://github.com/bitcoin/bitcoin/pull/26705#issuecomment-1353293405.

To test this PR, consider applying a diff as follows:
```diff
--- a/src/.clang-tidy
+++ b/src/.clang-tidy
@@ -1,16 +1,7 @@
 Checks: '
 -*,
-bugprone-argument-comment,
-bugprone-use-after-move,
-misc-unused-using-decls,
-modernize-use-default-member-init,
-modernize-use-nullptr,
-performance-for-range-copy,
 performance-move-const-arg,
 performance-no-automatic-move,
-performance-unnecessary-copy-initialization,
-readability-redundant-declaration,
-readability-redundant-string-init,
 '
 WarningsAsErrors: '
 bugprone-argument-comment,
@@ -28,4 +19,4 @@ readability-redundant-string-init,
 CheckOptions:
  - key: performance-move-const-arg.CheckTriviallyCopyableMove
    value: false
-HeaderFilterRegex: './qt'
+HeaderFilterRegex: '.'
```

MarcoFalke commented on src/rpc/util.h:
```diff
@@ -293,11 +293,11 @@ struct RPCResult {
         : RPCResult{cond, type, m_key_name, false, description, inner} {}
```
here as well? See also fab680ab9061a2dde0c4882caf002b057c1db8c6 from https://github.com/bitcoin/bitcoin/pull/26706

hebasto replied:
fab680ab9061a2dde0c4882caf002b057c1db8c6 has a wider effect as it adds `std::move` as well

Here, clang-tidy warnings/errors are fixed only. And changes are pretty focused and self-contained.

I'm not sure how to proceed further. Any suggestions?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26707#issuecomment-1385313594).

MarcoFalke replied:
If you rebase directly on fab92a5a5ab45227dbb0aa6d2cf4557b157b17e6, you can avoid changing the commit hash and (potentially) wasting review resources.

hebasto replied:
> If you rebase directly on [fab92a5](https://github.com/bitcoin/bitcoin/commit/fab92a5a5ab45227dbb0aa6d2cf4557b157b17e6), you can avoid changing the commit hash and (potentially) wasting review resources.

Done.

hebasto commented:
Updated f1f7472c2f0304daa83685b8ab7d0127d03f7ad7 -> 22eb1b449afb93eaa255dc00e28efc203ead7527 ([pr26707.01](https://github.com/hebasto/bitcoin/commits/pr26707.01) -> [pr26707.02](https://github.com/hebasto/bitcoin/commits/pr26707.02)):

- rebased
- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26707#discussion_r1050119855)

fanquake commented on src/script/miniscript.h:
```diff
@@ -1378,7 +1378,7 @@ inline NodeRef<Key> Parse(Span<const char> in, const Ctx& ctx)
     assert(constructed.size() == 1);
     assert(constructed[0]->ScriptSize() == script_size);
     if (in.size() > 0) return {};
-    const NodeRef<Key> tl_node = std::move(constructed.front());
```
cc @sipa @darosior

darosior replied:
TIL. My understanding is that without this patch `tl_node` would not be moved a couple lines below when returning from the function. Making it not `const` fixes this.

If my understanding is correct, ACK.

hebasto replied:
To be precise, this patch makes a compiler to call a move-constructor instead of a copy-constructor in the `return` statement.

MarcoFalke replied:
`NodeRef` is `std::shared_ptr<bla>`, so this shouldn't change the compiled binary except for a removed `std::atomic` add and sub.

hebasto committed:
clang-tidy: Fix `performance-move-const-arg` in headers

See https://clang.llvm.org/extra/clang-tidy/checks/performance/move-const-arg.html

hebasto committed:
clang-tidy: Fix `performance-no-automatic-move` in headers

See https://clang.llvm.org/extra/clang-tidy/checks/performance/no-automatic-move.html

MarcoFalke commented:
The diff in https://github.com/bitcoin/bitcoin/pull/26707#issue-1499057294 does not apply

hebasto commented:
Rebased  70e195ac9ff545fa327e6dc21e0be9bb7b768f81 -> 1308b837dc3499896ca73eafa51ac69b455cef00 ([pr26707.03](https://github.com/hebasto/bitcoin/commits/pr26707.03) -> [pr26707.04](https://github.com/hebasto/bitcoin/commits/pr26707.04)) on top of the merged #26706.

fanquake reviewed (approved):
ACK 1308b837dc3499896ca73eafa51ac69b455cef00
